chain/network/src/concurrency/scope/mod.rs

```diff
@@ -57,254 +57,245 @@ use crate::concurrency::signal;
-        let subscope = self.new_subscope();
-        let service = Service(Arc::downgrade(&subscope));
-        // Spawn a guard task in the Service which will prevent termination of the Service
-        // until the context is not canceled. See `Service` for a list
-        // of events canceling the Service.
-        Inner::spawn(subscope, async move { Ok(ctx::canceled().await) });
+    pub fn new_service<E2: 'static + Send>(self: Arc<Self>) -> Service<E2> {
+        let sub = Arc::new(TerminateGuard::new(&self.0.ctx));
+        let service = Service(Arc::downgrade(&sub), sub.0.clone());
+        // Spawn a guard task in `self` scope, so that it is not terminated
+        // before `sub` scope is.
+        TerminateGuard::spawn(self, async move {
+            let sub_inner = sub.0.clone();
+            // Spawn a guard task in `sub` scope, so that it is not terminated
+            // until its context is not canceled. See `Service` for a list
+            // of events canceling the Service.
+            TerminateGuard::spawn(sub, async move { Ok(ctx::canceled().await) });
+            sub_inner.terminated.recv().await;
+            Ok(())
+        });
         service
     }
 }
 
-/// Error returned when the `Service` has been already terminated
-/// and therefore spawning a task/service in it is not possible.
-#[derive(thiserror::Error, Debug)]
-#[error("ErrTerminated")]
+/// Error returned `Service::spawn()` and `Service::new_service()`
+/// when spawning new things on the service scope is not allowed, because the
+/// service has been already terminated. Returned by `JoinHandle::join` when
+/// the task has returned an error and therefore the service has been terminated.
+#[derive(thiserror::Error, Debug, Clone, Copy, PartialEq, Eq)]
+#[error("service has been terminated")]
 pub struct ErrTerminated;
 
 /// A service is a subscope which doesn't keep the scope
 /// alive, i.e. if all tasks spawned via `Scope::spawn` complete, the scope will
 /// be cancelled (even though tasks in a service may be still running).
 ///
 /// Note however that the scope won't be terminated until the tasks of the service complete.
-/// Service is cancelled when the handle is dropped, so make sure to store it somewhere.
-/// Service is cancelled when ANY of the tasks/services in the service returns an error.
-/// Service is cancelled when the parent scope/service is cancelled.
+/// Service is cancelled when ANY of the task in the service returns an error.
+/// Service is cancelled when the parent context is cancelled.
+/// Service is cancelled when `Service::terminate()` is called.
 /// Service is NOT cancelled just when all tasks within the service complete - in particular
 /// a newly started service has no tasks.
 /// Service is terminated when it is cancelled AND all tasks within the service complete.
-pub struct Service<E: 'static>(Weak<Inner<E>>);
+pub struct Service<E: 'static>(Weak<TerminateGuard<E>>, Arc<Inner<E>>);
```

Comment:
> So does it make sense to split Inner into two, each fulfilling one purpose?

It is split, just at different point. You ScopeReceipt is my Inner. I am moving back and forth using the oneshot::Receiver. Now it is infeasible, because Service error can be observed from multiple places.